Remove parallel node processing in PreFilter stage in volumerestrictions scheduler plugin

yibozhuang opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This change creates a `StorageInfoLister` interface
and have it under scheduler `SharedLister`.

The initial `StorageInfoLister` interface has a
`IsPVCUsedByPods` API which returns true/false depending on
whether the PVC (keyed by namespace/name) has at least
one scheduled pod using it.

In snapshot real implementation, add a `usedPVCSet`
key by PVC namespace/name which contains all PVCs
that have at least one scheduled pod using it.
During snapshot update, populate this set based on
whether the PVCRefCounts map for node(s) have been
updated since last snapshot.

With the `usedPVCSet` in snapshot, we can now determine
whether a PVC with ReadWriteOncePod access mode is being
used by another scheduled pod to achieve O(1) look up in
PreFilter and avoid needing to parallel process all nodes in the
`volumerestrictions` scheduler plugin.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #109687 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

yibozhuang commented:
/assign @Huang-Wei

denkensk commented on pkg/scheduler/internal/cache/cache.go:
```diff
@@ -259,6 +263,16 @@ func (cache *cacheImpl) UpdateSnapshot(nodeSnapshot *Snapshot) error {
 		cache.updateNodeInfoSnapshotList(nodeSnapshot, updateAllLists)
 	}
 
+	if updateUsedPVCSet {
```
Can we put this in `updateNodeInfoSnapshotList`?  No need to iterate over nodes multiple times.
https://github.com/yibozhuang/kubernetes/blob/7feaacd27ad18830ad8d299f6078d120526665f0/pkg/scheduler/internal/cache/cache.go#L292

denkensk commented on pkg/scheduler/internal/cache/snapshot.go:
```diff
@@ -98,6 +103,25 @@ func createNodeInfoMap(pods []*v1.Pod, nodes []*v1.Node) map[string]*framework.N
 	return nodeNameToInfo
 }
 
+func createUsedPVCs(pods []*v1.Pod) map[string]struct{} {
```
```suggestion
func createUsedPVCSet(pods []*v1.Pod) map[string]struct{} {
```

yibozhuang replied:
Thanks, fixed.

denkensk commented on pkg/scheduler/internal/cache/snapshot_test.go:
```diff
@@ -180,3 +181,48 @@ func TestCreateImageExistenceMap(t *testing.T) {
+	}
+
+	for i, test := range tests {
+		t.Run(fmt.Sprintf("case_%d", i), func(t *testing.T) {
```
name may be better than `case_id`

yibozhuang replied:
Sounds good, added a name field.

denkensk commented:
/ok-to-test

Huang-Wei commented on pkg/scheduler/internal/cache/cache.go:
```diff
@@ -220,6 +223,7 @@ func (cache *cacheImpl) UpdateSnapshot(nodeSnapshot *Snapshot) error {
 			// all the nodes are updated before the existing snapshot. We are done.
 			break
 		}
+		updateUsedPVCSet = true
```
Can we make it conditional, with a minimum calculation cost like `updateNodesHavePodsWithAffinity`?

yibozhuang replied:
Sure, added an additional check
```go
if !reflect.DeepEqual(existing.PVCRefCounts, clone.PVCRefCounts) {
	updateUsedPVCSet = true
}
```
in `UpdateSnapshot()`

and in `types.go` when we clone `NodeInfo`, updated to do a proper clone of the `PVCRefCounts` map rather than updating the reference so that we can perform this comparison when taking snapshot.

yibozhuang replied:
Sure, added an additional check
```go
if !reflect.DeepEqual(existing.PVCRefCounts, clone.PVCRefCounts) {
	updateUsedPVCSet = true
}
```
in `UpdateSnapshot()`

and in `types.go` when we clone `NodeInfo`, updated to do a proper clone of the `PVCRefCounts` map rather than updating the reference so that we can perform this comparison when taking snapshot.

yibozhuang commented on pkg/scheduler/internal/cache/snapshot.go:
```diff
@@ -36,7 +36,10 @@ type Snapshot struct {
-	generation                                   int64
+	// usedPVCSet contains a set of PVC names that have one or more scheduled pods using them,
+	// keyed in the format "namespace/name".
+	usedPVCSet map[string]struct{}
```
Sure thing, done!

yibozhuang commented on pkg/scheduler/internal/cache/snapshot.go:
```diff
@@ -98,6 +103,25 @@ func createNodeInfoMap(pods []*v1.Pod, nodes []*v1.Node) map[string]*framework.N
+				continue
+			}
+
+			key := fmt.Sprintf("%s/%s", pod.Namespace, v.PersistentVolumeClaim.ClaimName)
```
Sure thing, created a `GetNamespacedName()` under framework `types.go` and have `updatePVCRefCounts`, `createUsedPVCSet`, and `volumerestrictions` plugin consume it.

Huang-Wei commented on pkg/scheduler/framework/types.go:
```diff
@@ -571,6 +571,11 @@ func (n *NodeInfo) Clone() *NodeInfo {
 	if len(n.PodsWithRequiredAntiAffinity) > 0 {
 		clone.PodsWithRequiredAntiAffinity = append([]*PodInfo(nil), n.PodsWithRequiredAntiAffinity...)
 	}
+	if len(n.PVCRefCounts) > 0 {
```
The `if` is redundant.

yibozhuang replied:
Thanks, fixed.

Huang-Wei commented on pkg/scheduler/internal/cache/cache.go:
```diff
@@ -237,6 +241,9 @@ func (cache *cacheImpl) UpdateSnapshot(nodeSnapshot *Snapshot) error {
 			if (len(existing.PodsWithRequiredAntiAffinity) > 0) != (len(clone.PodsWithRequiredAntiAffinity) > 0) {
 				updateNodesHavePodsWithRequiredAntiAffinity = true
 			}
+			if !reflect.DeepEqual(existing.PVCRefCounts, clone.PVCRefCounts) {
```
I think we should (1) check if updateUsedPVCSet is already false, and (2) just verify the entires are identical (the int value doesn't quite matter)


```go
			if !updateUsedPVCSet {
				if len(existing.PVCRefCounts) != len(clone.PVCRefCounts) {
					updateUsedPVCSet = true
				} else {
					for pvcKey := range clone.PVCRefCounts {
						if _, ok := existing.PVCRefCounts[pvcKey]; !ok {
							updateUsedPVCSet = true
							break
						}
					}
				}
			}
```

yibozhuang replied:
Sounds good, fixed.

Huang-Wei commented on pkg/scheduler/internal/cache/cache_test.go:
```diff
@@ -1239,6 +1240,32 @@ func TestSchedulerCache_UpdateSnapshot(t *testing.T) {
 	}
 
+	// Add a few of pods with PVC, starting on the third node.
+	podsWithPVC := []*v1.Pod{}
```
It's more conventional:

```suggestion
	var podsWithPVC []*v1.Pod
```

yibozhuang replied:
Sure, and fixed in all the other places in this file.

Huang-Wei commented on pkg/scheduler/internal/cache/cache_test.go:
```diff
@@ -1239,6 +1240,32 @@ func TestSchedulerCache_UpdateSnapshot(t *testing.T) {
+						Name: fmt.Sprintf("test-pvc%v", i),
+						VolumeSource: v1.VolumeSource{
+							PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+								ClaimName: fmt.Sprintf("test-pvc%v", i),
```
Do you want to test the case multiple pods shared a PVC? If so, `i%2` or `i%4` here.

yibozhuang replied:
Changed to `i%4` here.

Huang-Wei commented on pkg/scheduler/internal/cache/cache_test.go:
```diff
@@ -1476,6 +1571,11 @@ func TestSchedulerCache_UpdateSnapshot(t *testing.T) {
 			}
 
+			// Compare content of the used PVC set
+			if !test.expectedUsedPVCSet.Equal(snapshot.usedPVCSet) {
```
let's use cmp.Diff():

```go
if diff := cmp.Diff(want, got); diff != "" {
    t.Errorf("Unexpected usedPVCSet (-want +got):\n%s", diff)
}
```

yibozhuang replied:
Sure thing, fixed.

Huang-Wei commented on pkg/scheduler/internal/cache/cache.go:
```diff
@@ -237,6 +240,18 @@ func (cache *cacheImpl) UpdateSnapshot(nodeSnapshot *Snapshot) error {
+				if len(existing.PVCRefCounts) != len(clone.PVCRefCounts) {
+					updateUsedPVCSet = true
+				} else {
+					for pvcKey := range clone.PVCRefCounts {
+						if _, found := existing.PVCRefCounts[pvcKey]; !found {
+							updateUsedPVCSet = true
+							break
+						}
+					}
```
Can you make this part UTed?

yibozhuang replied:
Sure thing added two UT covering:
- same ref count length, same content
- same ref count length, different content

Huang-Wei replied:
I suppose L249 is still not hit. How about building a case like this:

- a pod with PVC0 lands on node0
- updateSnapshot
- delete the pod
- another pod with PVC1 lands on node0
- updateSnapshot

In this case, we should be able to cover L249.

yibozhuang replied:
Sure thing
```
Add then Remove pod with PVC
```
test case covers ^ and confirmed the block is fully covered with go test coverage

yibozhuang commented on pkg/scheduler/internal/cache/snapshot_test.go:
```diff
@@ -180,3 +181,50 @@ func TestCreateImageExistenceMap(t *testing.T) {
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			usedPVCs := createUsedPVCSet(test.pods)
+			if !test.expected.Equal(usedPVCs) {
```
Sure thing, done.

yibozhuang commented:
/retest

Huang-Wei commented on pkg/scheduler/internal/cache/cache_test.go:
```diff
@@ -1239,6 +1241,32 @@ func TestSchedulerCache_UpdateSnapshot(t *testing.T) {
 
+	// Add a few of pods with PVC, starting on the third node.
+	var podsWithPVC []*v1.Pod
+	for i := 2; i < 8; i++ {
```
to distiinguish with podWithAffinity, you can build them in a declarative manner:

```
podsWithAffinity := []*v1.Pod{
    st.MakePod().Name("p-affinity-0").Node("node0")...,
}

podsWithPVC := []*v1.Pod{
    st.MakePod().Name("p-pvc-0").Node("node0")...,
    ....
}
```

Then we don't need to start with index 2, which makes the test hard to read.

yibozhuang replied:
Thanks, done!

yibozhuang committed:
Adding StorageInfoLister to SharedLister

This change creates a StorageInfoLister interface
and have it under scheduler SharedLister.

The initial StorageInfoLister interface has a
IsPVCUsedByPods which returns true/false depending on
whether the PVC (keyed by namespace/name) has at least
one scheduled pod using it.

In snapshot real implementation, add a usedPVCSet
key by PVC namespace/name which contains all PVCs
that have at least one scheduled pod using it.
During snapshot update, populate this set based on
whether the PVCRefCounts map for node(s) have been
updated since last snapshot.

Signed-off-by: Yibo Zhuang <yibzhuang@gmail.com>

yibozhuang committed:
Remove parallel node processing in volumerestrictions

Leverage the usedPVCSet in snapshot to determine
whether a PVC with ReadWriteOncePod access mode is being
used by another scheduled pod to achieve O(1) look up in
PreFilter and avoid needing to parallel process all nodes.

Signed-off-by: Yibo Zhuang <yibzhuang@gmail.com>

Huang-Wei commented:
Thanks @yibozhuang !

/lgtm
/approve

MaciekPytel commented:
The SharedLister is an interface between scheduler and Cluster Autoscaler. CA uses its own [implementation of SharedLister](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/simulator/cluster_snapshot.go#L29) as a way to run Filters on in-memory NodeInfos that represent potential nodes that may be created by autoscaler. 

Moving PV-related logic into SharedLister could in principle help solve longstanding issues around autoscaling and pods using PVs (if all access to PV/PVC would actually go through the SharedLister), but we should make sure that any changes we make here are done with co-ordination with sig-autoscaling to:
 * Make sure the design is compatible with CA;
 * Necessary changes are also implemented to CA (as-is this PR will cause CA to stop compiling, as its ClusterSnapshot no longer satisfies the new interface).

@ahg-g @alculquicondor @mwielgus @x13n

Huang-Wei commented:
@MaciekPytel apologize for not having got CA involved, which causes the compiling issue.

Furthermore, does CA run **PreFilter** and Filter to simulate the scheduling process? If so, this PR should be compatible with CA's design, and we can work on a CA PR to fix the compiling issue.

x13n commented:
It runs both: https://github.com/kubernetes/autoscaler/blob/e3ad872747917613f8c132cf22c7cd9b0eca9915/cluster-autoscaler/simulator/scheduler_based_predicates_checker.go#L106-L122

The problem is elsewhere though, there is an internal interface in CA which extends SharedLister: https://github.com/kubernetes/autoscaler/blob/e3ad872747917613f8c132cf22c7cd9b0eca9915/cluster-autoscaler/simulator/cluster_snapshot.go#L29

We now need to update implementations of SharedLister interface to keep the code compiling:
- https://github.com/kubernetes/autoscaler/blob/e3ad872747917613f8c132cf22c7cd9b0eca9915/cluster-autoscaler/simulator/basic_cluster_snapshot.go
- https://github.com/kubernetes/autoscaler/blob/e3ad872747917613f8c132cf22c7cd9b0eca9915/cluster-autoscaler/simulator/delta_cluster_snapshot.go
- https://github.com/kubernetes/autoscaler/blob/e3ad872747917613f8c132cf22c7cd9b0eca9915/cluster-autoscaler/simulator/delegating_shared_lister.go

ahg-g commented:
@Huang-Wei for the future can we please have more than one lead look at such major changes to get consensus?

jayantjain93 commented on pkg/scheduler/framework/listers.go:
```diff
@@ -18,17 +18,25 @@ package framework
+type StorageInfoLister interface {
+	// IsPVCUsedByPods returns true/false on whether the PVC is used by one or more scheduled pods,
+	// keyed in the format "namespace/name".
+	IsPVCUsedByPods(key string) bool
```
@ahg-g , I think the func should implement an error return as well.

Do you think we can add that to the next release ?
